PXC-3580: Aggressive network outages on one node makes the whole cluster unusable

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3580

node_1 - IST sender
node_2 - IST receiver

Problem on the donor side:
node_2 partitioned after receiving IST, but before processing all WSs
and signaling node_1 that it finished.

1. node_2 partitioned
2. node_1 tries to reconnect to node_2, and while it is not possible
within evs.suspect_timeout/evs.inactive_timeout, the node is declared as
inactive
3. node_2 becomes available again
4. node_2 requests for IST
5. node_1 starts AsyncSender (it locks GCache, which will be unlocked
when AsyncSender finishes), node_2 starts Receiver.
6. IST is being sent node_1 -> node_2. node_2 buffers all writesets in
the queue which is processed by node_2's replication threads.
8. There is no SST so node_1 shifts its state DONOR -> JOINED -> SYNCED
and processes process_commit_cut seq() which triggers async action
(ServiceThd) of release_seqno(). It cannot be released, as AsyncSender
is having GCache lock.
7. All WSs are transferred, however AsycSender waits for Receiver to
close the connection (AsyncSender::send() => send_eof()) holding GCache
lock.
9. node_2 partitioned again
10. node_1 tries to reconnect again and node_2 gets removed from the
cluster
11. CC is processed by node_1. Certification position
is being adjusted and ServiceThd::flush is called. It blocks, as we
already have pending release_seqno().
12. process_conf_change enters local_monitor_ with CC seqno
13. ReplicatorSMM::process_prim_conf_change() is called, certification
position is adjusted. GCache release_seqno is requested (async action
done by service_thd_). However, this action is blocked, because GCache
is locked by IST AsyncSender that was for node_2. So we wait on flush,
having local_monitor_ entered.
14. Query from the client is executed. It gets into the certification
stage and attempts to enter local_monitor_
(enter_local_monitor_for_cert), but is blocked.

Solution:
When node_2 is being partitioned, AsyncSender associated with this node
should be unconditionally terminated.

Problem on the joiner side:
1. on node_2 Receiver::run() uses synchronous socket reads to get WSs
from donor.
2. node_2 partitioned
3. node_1 detects the problem, closes AsyncSender
4. node_2 waits infinitely for data from node_1

Solution:
It would be better to use async socket read together with a timer.
Unfortunately, we are not using boost asio. Additionally, it would need
deeper refactoring, testing.
Instead of huge changes, SocketWatchdog was introduced which
terminates the socket (and the whole Receiver) if no data is received
within 10 seconds.
In such a case node_2 restart is required, but this is yet another
existing issue (by design) that if node_2 didn't receive all WSs
expected, it self-leaves the cluster with an appropriate message in
the error log.

kamil-holubicki committed:
PXC-3580: Aggressive network outages on one node
makes the whole cluster unusable

https://jira.percona.com/browse/PXC-3580

node_1 - IST sender
node_2 - IST receiver

https://jira.percona.com/browse/PXC-3580

node_1 - IST sender
node_2 - IST receiver

Problem on the donor side:
node_2 partitioned after receiving IST, but before processing all WSs
and signaling node_1 that it finished.

1. node_2 partitioned
2. node_1 tries to reconnect to node_2, and while it is not possible
within evs.suspect_timeout/evs.inactive_timeout, the node is declared as
inactive
3. node_2 becomes available again
4. node_2 requests for IST
5. node_1 starts AsyncSender (it locks GCache, which will be unlocked
when AsyncSender finishes), node_2 starts Receiver.
6. IST is being sent node_1 -> node_2. node_2 buffers all writesets in
the queue which is processed by node_2's replication threads.
8. There is no SST so node_1 shifts its state DONOR -> JOINED -> SYNCED
and processes process_commit_cut seq() which triggers async action
(ServiceThd) of release_seqno(). It cannot be released, as AsyncSender
is having GCache lock.
7. All WSs are transferred, however AsycSender waits for Receiver to
close the connection (AsyncSender::send() => send_eof()) holding GCache
lock.
9. node_2 partitioned again
10. node_1 tries to reconnect again and node_2 gets removed from the
cluster
11. CC is processed by node_1. Certification position
is being adjusted and ServiceThd::flush is called. It blocks, as we
already have pending release_seqno().
12. process_conf_change enters local_monitor_ with CC seqno
13. ReplicatorSMM::process_prim_conf_change() is called, certification
position is adjusted. GCache release_seqno is requested (async action
done by service_thd_). However, this action is blocked, because GCache
is locked by IST AsyncSender that was for node_2. So we wait on flush,
having local_monitor_ entered.
14. Query from the client is executed. It gets into the certification
stage and attempts to enter local_monitor_
(enter_local_monitor_for_cert), but is blocked.

Solution:
When node_2 is being partitioned, AsyncSender associated with this node
should be unconditionally terminated.

Problem on the joiner side:
1. on node_2 Receiver::run() uses synchronous socket reads to get WSs
from donor.
2. node_2 partitioned
3. node_1 detects the problem, closes AsyncSender
4. node_2 waits infinitely for data from node_1

Solution:
It would be better to use async socket read together with a timer.
Unfortunately, we are not using boost asio. Additionally, it would need
deeper refactoring, testing.
Instead of huge changes, SocketWatchdog was introduced which
terminates the socket (and the whole Receiver) if no data is received
within 10 seconds.
In such a case node_2 restart is required, but this is yet another
existing issue (by design) that if node_2 didn't receive all WSs
expected, it self-leaves the cluster with an appropriate message in
the error log.

kamil-holubicki commented:
https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/224
Failures caused by unstable tests.
galera.galera_alter_table_big is the permanent HEAD failure.

dutow commented on galera/src/ist.cpp:
```diff
@@ -472,223 +562,236 @@ void galera::ist::Receiver::run()
-
-            gcs_action& act(ret.first);
+            SocketWatchdog watchdog([&ssl_stream, &socket, this]() {
+                fprintf(stderr, "SocketWatchdog expired. use_ssl_: %d\n", use_ssl_);
```
Shouldn't this print with a normal galera logging macro?

kamil-holubicki replied:
fixed

dutow commented on galera/src/replicator_smm.cpp:
```diff
@@ -2631,11 +2631,21 @@ galera::ReplicatorSMM::process_conf_change(void*                    recv_ctx,
 
 void galera::ReplicatorSMM::drain_monitors_for_local_conf_change(const gcs_act_cchange& cc)
 {
+   std::vector<std::string> allowed_IST_clients;
```
you could reserve / allocate a cc.memb.size() vector here.

kamil-holubicki replied:
fixed

dutow commented on galera/src/ist.cpp:
```diff
@@ -395,6 +403,88 @@ galera::ist::Receiver::prepare(wsrep_seqno_t const first_seqno,
+
+        }
+
+        void restart() {
```
I'm not sure if the name of this function is good: if the loop is currently running, it won't restart it, only if it is currently stopped.

While unlikely, it also seems possible that a quick stop(); restart(); sequence will happen without the thread noticing it, continuing the current counter instead of restarting it.

kamil-holubicki replied:
fixed

dutow reviewed (commented):
Only some minor comments, otherwise it looks good.

kamil-holubicki committed:
PXC-3580: Aggressive network outages on one node
makes the whole cluster unusable

https://jira.percona.com/browse/PXC-3580

Addressed review comments.

Additionally added assert in
ReplicatorSMM::drain_monitors_for_local_conf_change() which was removed
by omission.
